### PR TITLE
module compression is now again settable via var

### LINKED ISSUE #83571: ansible_module_compression is not available since 2.16.0b1
### Summary

With #80647 the feature to be able to change module compression *per-host* was removed.

Previously, inventory script was able to set `ansible_module_compression=ZIP_STORED` for some host. This is mandatory for my use case. A same playbook is managing both high-end servers and low-end (as in 64MB RAM) machines, with custom compilation toolchain, so Python has no zlib.

Moreover, this was mentioned as "minor" change. Maybe because it was never documented officially? It's still visible in a lot of "example" `ansible.cfg`, but not from ansible repository.

### Issue Type

Bug Report

### Component Name

core

### Ansible Version

```console
$ ansible --version
ansible [core 2.16.8]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /export/gls/backups/venv-ansible-py311/lib64/python3.11/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = ./bin/ansible
  python version = 3.11.9 (main, Apr 18 2024, 16:44:43) [GCC] (/export/gls/backups/venv-ansible-py311/bin/python3.11)
  jinja version = 3.1.4
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
ANSIBLE_PIPELINING(/export/gls/backups/ansible_unitronics/ansible.cfg) = True
COLOR_VERBOSE(/export/gls/backups/ansible_unitronics/ansible.cfg) = bright blue
CONFIG_FILE() = /export/gls/backups/ansible_unitronics/ansible.cfg
DEFAULT_FORKS(/export/gls/backups/ansible_unitronics/ansible.cfg) = 50
DEFAULT_HOST_LIST(/export/gls/backups/ansible_unitronics/ansible.cfg) = ['/export/gls/backups/ansible_unitronics/inventory']
DEFAULT_INTERNAL_POLL_INTERVAL(/export/gls/backups/ansible_unitronics/ansible.cfg) = 0.1
DEFAULT_JINJA2_EXTENSIONS(/export/gls/backups/ansible_unitronics/ansible.cfg) = jinja2.ext.do
DEFAULT_MODULE_PATH(/export/gls/backups/ansible_unitronics/ansible.cfg) = ['/export/gls/backups/ansible_unitronics/library']
DEFAULT_VAULT_PASSWORD_FILE(/export/gls/backups/ansible_unitronics/ansible.cfg) = /export/gls/backups/ansible_unitronics/ansible-password
HOST_KEY_CHECKING(/export/gls/backups/ansible_unitronics/ansible.cfg) = False
PAGER(env: PAGER) = less
SHOW_CUSTOM_STATS(/export/gls/backups/ansible_unitronics/ansible.cfg) = True

CALLBACK:
========

default:
_______
show_custom_stats(/export/gls/backups/ansible_unitronics/ansible.cfg) = True

CONNECTION:
==========

local:
_____
pipelining(/export/gls/backups/ansible_unitronics/ansible.cfg) = True

paramiko_ssh:
____________
host_key_checking(/export/gls/backups/ansible_unitronics/ansible.cfg) = False
ssh_args(/export/gls/backups/ansible_unitronics/ansible.cfg) = -q -o Compression=yes -o ControlMaster=auto -o ControlPersist=5m -o Check>

psrp:
____
pipelining(/export/gls/backups/ansible_unitronics/ansible.cfg) = True

ssh:
___
control_path(/export/gls/backups/ansible_unitronics/ansible.cfg) = %(directory)s/ansible-ssh-%%h-%%r
host_key_checking(/export/gls/backups/ansible_unitronics/ansible.cfg) = False
pipelining(/export/gls/backups/ansible_unitronics/ansible.cfg) = True
ssh_args(/export/gls/backups/ansible_unitronics/ansible.cfg) = -q -o Compression=yes -o ControlMaster=auto -o ControlPersist=5m -o Check>

winrm:
_____
pipelining(/export/gls/backups/ansible_unitronics/ansible.cfg) = True
```

### OS / Environment

SuSE Linux Enterprise 15 SP5

### Steps to Reproduce

```yaml (paste below)

```
Use an inventory host with `ansible_module_compression=ZIP_STORED`

### Expected Results

I expect the module to be transferred without compression for this host.

### Actual Results

```console
# ansible -m ping 10.37.91.20 -vvvv
ansible [core 2.16.8]
  config file = /export/backups/ansible_unitronics/ansible.cfg
  configured module search path = ['/export/backups/ansible_unitronics/library']
  ansible python module location = /export/backups/venv-ansible-py311/lib64/python3.11/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /export/backups/venv-ansible-py311/bin/ansible
  python version = 3.11.9 (main, Apr 18 2024, 16:44:43) [GCC] (/export/backups/venv-ansible-py311/bin/python3.11)
  jinja version = 3.1.4
  libyaml = True
Using /export/backups/ansible_unitronics/ansible.cfg as config file
setting up inventory plugins
Loading collection ansible.builtin from 
host_list declined parsing /export/backups/ansible_unitronics/inventory/inventory.py as it did not pass its verify_file() method
Not replacing invalid character(s) "{'-'}" in group name (panel-pc)
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details
Parsed /export/backups/ansible_unitronics/inventory/inventory.py inventory source with script plugin
setting up inventory plugins
host_list declined parsing /export/backups/ansible_unitronics/inventory/static as it did not pass its verify_file() method
script declined parsing /export/backups/ansible_unitronics/inventory/static as it did not pass its verify_file() method
auto declined parsing /export/backups/ansible_unitronics/inventory/static as it did not pass its verify_file() method
Set default localhost to localhost
Parsed /export/backups/ansible_unitronics/inventory/static inventory source with ini plugin
Loading callback plugin minimal of type stdout, v2.0 from /export/backups/venv-ansible-py311/lib64/python3.11/site-packages/ansible/plugins/callback/minimal.py
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.
<10.37.91.20> Attempting python interpreter discovery
<10.37.91.20> ESTABLISH SSH CONNECTION FOR USER: root
<10.37.91.20> SSH: EXEC ssh -vvv -q -o Compression=yes -o ControlMaster=auto -o ControlPersist=5m -o CheckHostIP=no -o UserKnownHostsFile=/dev/null -o ForwardAgent=yes -o StrictHostKeyChecking=no -o 'IdentityFile="id_backup"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/ansible-ssh-%h-%r"' 10.37.91.20 '/bin/sh -c '"'"'echo PLATFORM; uname; echo FOUND; command -v '"'"'"'"'"'"'"'"'python3.12'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'python3.11'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'python3.10'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'python3.9'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'python3.8'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'python3.7'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'python3.6'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'/usr/bin/python3'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'/usr/libexec/platform-python'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'python2.7'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'/usr/bin/python'"'"'"'"'"'"'"'"'; command -v '"'"'"'"'"'"'"'"'python'"'"'"'"'"'"'"'"'; echo ENDFOUND && sleep 0'"'"''
<10.37.91.20> (0, b'PLATFORM\nLinux\nFOUND\n/usr/bin/python2.7\n/usr/bin/python\n/usr/bin/python\nENDFOUND\n', b'')
<10.37.91.20> ESTABLISH SSH CONNECTION FOR USER: root
<10.37.91.20> SSH: EXEC ssh -vvv -q -o Compression=yes -o ControlMaster=auto -o ControlPersist=5m -o CheckHostIP=no -o UserKnownHostsFile=/dev/null -o ForwardAgent=yes -o StrictHostKeyChecking=no -o 'IdentityFile="id_backup"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/ansible-ssh-%h-%r"' 10.37.91.20 '/bin/sh -c '"'"'/usr/bin/python2.7 && sleep 0'"'"''
<10.37.91.20> (0, b'{"osrelease_content": "NAME=AxotecLinux\\nVERSION=G02.V08\\nID=axoteclinux\\nVERSION_ID=G02.V08\\nPRETTY_NAME=\\"AxotecLinux G02.V08\\"\\n", "platform_dist_result": ["", "", ""]}\n', b'')
<10.37.91.20> Python interpreter discovery fallback (unsupported Linux distribution: axoteclinux)
Using module file /export/backups/venv-ansible-py311/lib64/python3.11/site-packages/ansible/modules/ping.py
Pipelining is enabled.
<10.37.91.20> ESTABLISH SSH CONNECTION FOR USER: root
<10.37.91.20> SSH: EXEC ssh -vvv -q -o Compression=yes -o ControlMaster=auto -o ControlPersist=5m -o CheckHostIP=no -o UserKnownHostsFile=/dev/null -o ForwardAgent=yes -o StrictHostKeyChecking=no -o 'IdentityFile="id_backup"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/ansible-ssh-%h-%r"' 10.37.91.20 '/bin/sh -c '"'"'/usr/bin/python2.7 && sleep 0'"'"''
<10.37.91.20> (1, b'', b'Traceback (most recent call last):\n  File "<stdin>", line 107, in <module>\n  File "<stdin>", line 99, in _ansiballz_main\n  File "<stdin>", line 44, in invoke_module\nzipimport.ZipImportError: can\'t decompress data; zlib not available\n')
<10.37.91.20> Failed to connect to the host via ssh: Traceback (most recent call last):
  File "<stdin>", line 107, in <module>
  File "<stdin>", line 99, in _ansiballz_main
  File "<stdin>", line 44, in invoke_module
zipimport.ZipImportError: can't decompress data; zlib not available
The full traceback is:
Traceback (most recent call last):
  File "<stdin>", line 107, in <module>
  File "<stdin>", line 99, in _ansiballz_main
  File "<stdin>", line 44, in invoke_module
zipimport.ZipImportError: can't decompress data; zlib not available
[WARNING]: Platform linux on host 10.37.91.20 is using the discovered Python interpreter at /usr/bin/python2.7, but future installation
of another Python interpreter could change the meaning of that path. See https://docs.ansible.com/ansible-
core/2.16/reference_appendices/interpreter_discovery.html for more information.
10.37.91.20 | FAILED! => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python2.7"
    },
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"<stdin>\", line 107, in <module>\n  File \"<stdin>\", line 99, in _ansiballz_main\n  File \"<stdin>\", line 44, in invoke_module\nzipimport.ZipImportError: can't decompress data; zlib not available\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by bcoca**
fixes #83571

##### ISSUE TYPE

- Bugfix Pull Request

### CODE REVIEW

File: lib/ansible/config/base.yml [hunk 1]
```diff
@@ -863,8 +863,8 @@ DEFAULT_MODULE_COMPRESSION:
   env: []
   ini:
   - {key: module_compression, section: defaults}
-# vars:
-#   - name: ansible_module_compression
+ vars:
+   - name: ansible_module_compression
```
Comment (Glandos) [inline lib/ansible/config/base.yml]:
```suggestion
  vars:
    - name: ansible_module_compression
```
It should be indented with one more space

File: changelogs/fragments/mc_fix.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+bufixes:
```
Comment (Akasurde) [inline changelogs/fragments/mc_fix.yml]:
```suggestion
bugfixes:
```

**Review by mkrizek — APPROVED** [review]

**Comment by Glandos** [comment]
Do you know if it can be backported to 2.16?

It's only personal, I have machine with python3.6 only (yeaaah, CentOS 7 is still living), and I can't go to 2.17 as far as I understood.

**Comment by bcoca** [comment]
I had already created backports #83591 for 2.17 and #83592 for 2.16

**Pull request merged**